internal/perf/perf.go

```diff
@@ -549,25 +549,54 @@ perfLoop:
 			break perfLoop
 		}
 
-		select {
-		case <-signalCh:
-			break perfLoop
-		case pr.bfr <- i:
-			i++
-			if time.Since(lastCheckedTime).Seconds() > pr.cfg.MaxTimePerAction.Seconds() {
+		if pr.cfg.MaxSubmissionsPerSecond > 0 {
+			// control send rate
+			secondTicker := time.NewTicker(1 * time.Second)
+			select {
+			case <-signalCh:
+				break perfLoop
+			case <-secondTicker.C:
+				for j := 0; j < pr.cfg.MaxSubmissionsPerSecond; j++ {
+					pr.bfr <- j
+				}
+				i++
+				if time.Since(lastCheckedTime).Seconds() > pr.cfg.MaxTimePerAction.Seconds() {
+					if pr.detectDelinquentMsgs() && pr.cfg.DelinquentAction == conf.DelinquentActionExit.String() {
+						break perfLoop
+					}
+					lastCheckedTime = time.Now()
+				}
+				break
+			case <-timeout:
 				if pr.detectDelinquentMsgs() && pr.cfg.DelinquentAction == conf.DelinquentActionExit.String() {
```

Comment:
Don't like that we are duplicating a lot of code but understand that a conditional case statement doesn't really exist. I think what we can do is switch the `pr.bfr <- i` into `default:` and then within that do either the check for the deliqunent and consume the ticker or directly `pr.bfr <- i`

Reply:
@EnriqueL8 
I'm not sure whether this comment is just for the non-timer-based trigger or the constant trigger branch. Therefore, I cannot tell how your counter-proposal will affect readability.

I feel maybe explaining it in code is easier in this sort of thing?

Reply:
Sure, let me try that

Reply:
I guess it's too convoluted to do what I suggested, was thinking the following sudo code:

```
select {
  case <-signalCh:
          break perfLoop
  case <-timeout:
          // same code
   default:
        if pr.cfg.MaxSubmissionsPerSecond != nil {
             secondTicker := time.NewTicker(1 * time.Second)
             
             // Feel like this could be changed to a sleep maybe
             select {
                        case <-secondTicker.C:
				for j := 0; j < pr.cfg.MaxSubmissionsPerSecond; j++ {
					pr.bfr <- j
				}    
             }
             break;
        }

       // Non Max submission        
        pr.bfr <- i

}
```

Reply:
Maybe the right tradeoff is we keep the structure of your code and just make the things inside the cases common code functions that just get called?

Reply:
ah, I see. I think we can switch to rate limiter maybe. let me give it a go. Thanks

Reply:
@EnriqueL8 ready for another look. Thanks for raising it

Reply:
nice taking a look